Userland/xargs.cpp

```diff
@@ -0,0 +1,292 @@
+#include <string.h>
+#include <sys/wait.h>
+
+bool run_command(Vector<char*>&& child_argv, bool verbose, bool is_stdin, int devnull_fd);
+
+enum Decision {
+    Unget,
+    Continue,
+    Stop,
+};
+bool read_items(FILE* fp, char entry_separator, Function<Decision(StringView)>);
+
+class ParsedInitialArguments {
+public:
+    ParsedInitialArguments(Vector<const char*>&, const StringView& placeholder);
+
+    void for_each_joined_argument(const StringView&, Function<void(const String&)>) const;
+
+    size_t size() const { return m_all_parts.size(); }
+
+private:
+    Vector<Vector<StringView>> m_all_parts;
+};
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio rpath proc exec", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    const char* placeholder = nullptr;
+    bool split_with_nulls = false;
+    const char* specified_delimiter = "\n";
+    Vector<const char*> arguments;
+    bool verbose = false;
+    const char* file_to_read = "-";
+    int max_lines_for_one_command = 0;
+    int max_bytes_for_one_command = ARG_MAX;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_option(placeholder, "Placeholder string to be replaced in arguments", "replace", 'I', "placeholder");
+    args_parser.add_option(split_with_nulls, "Split input items with the null character instead of newline", "null", '0');
+    args_parser.add_option(specified_delimiter, "Split the input items with the specified character", "delimiter", 'd', "delim");
+    args_parser.add_option(verbose, "Display each command before executing it", "verbose", 'v');
+    args_parser.add_option(file_to_read, "Read arguments from the specified file instead of stdin", "arg-file", 'a', "file");
+    args_parser.add_option(max_lines_for_one_command, "Use at most max-lines lines to create a command", "line-limit", 'L', "max-lines");
+    args_parser.add_option(max_bytes_for_one_command, "Use at most max-chars characters to create a command", "char-limit", 's', "max-chars");
+    args_parser.add_positional_argument(arguments, "Command and any initial arguments for it", "command", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    size_t max_bytes = min(ARG_MAX, max_bytes_for_one_command);
+    size_t max_lines = max(max_lines_for_one_command, 0);
+
+    if (!split_with_nulls && strlen(specified_delimiter) > 1) {
+        fprintf(stderr, "xargs: the delimiter must be a single byte\n");
+        return 1;
+    }
+
+    char entry_separator = split_with_nulls ? '\0' : *specified_delimiter;
+
+    StringView placeholder_view { placeholder };
+
+    if (!placeholder_view.is_empty())
+        max_lines = 1;
+
+    if (arguments.is_empty())
+        arguments.append("echo");
+
+    ParsedInitialArguments initial_arguments(arguments, placeholder_view);
+
+    FILE* fp = stdin;
+    bool is_stdin = true;
+
+    if (StringView { "-" } != file_to_read) {
+        // A file was specified, try to open it.
+        fp = fopen(file_to_read, "r");
```

Comment:
I believe you're still leaking it to the children. If you don't want to implement cloexec support for `fopen()`, just make sure to close it after forking.

Reply:
oh, I'll mess with fopen then
should I just do the gnu extension `e`?

Reply:
I would do that yeah.